students/timm67/session07/html_render.py

```diff
@@ -7,25 +7,201 @@
+    def __init__(self, content_in=None, **kwargs):
+        if content_in is not None:
+            self._content = [content_in]
+        else:
+            self._content = []
+
+        # save attributes passed in as kwargs
+        if len(kwargs.items()) > 0:
+            for key, value in kwargs.items():
+                self._attributes[key] = value
+        else:
+            self._attributes = {}
 
     def append(self, new_content):
-        self.content.append(new_content)
+        if new_content is not None:
+            self._content.append(new_content)
+
+    def render_attributes(self, out_fd, indent_in):
+        num_attrib = len(self._attributes)
+        if(num_attrib > 0):
+            out_fd.write('{0}<{1} '.format(indent_in, self._tag))
+            for key, value in self._attributes.items():
+                out_fd.write('{0}="{1}" '.format(key, value))
+        else:
+            out_fd.write('{0}<{1}>'.format(indent_in, self._tag))
+        return num_attrib
 
-    def render(self, out_file):
+    def render(self, out_fd, indent_in=''):
         try:
-            # TODO: may want to append to already opened file (i.e. call render multiple times)
-            with open(out_file, 'w') as fd:
-                for line in self.content:
-                    try:
-                        fd.write(line)
-                    except IOError:
-                        print("I/O Error with file [{0}] on writeline".format(out_file))
-                        return
+            # render DOCTYPE
+            if isinstance(self, Html) is True:
+                out_fd.write('{0}\n'.format(self._doctype))
+
+            num_attr = self.render_attributes(out_fd, indent_in)
+            if (num_attr > 0):
+                out_fd.write('>\n')
+            else:
+                out_fd.write('\n')
+
+            # loop through the list of contents:
+            for content_item in self._content:
+                try:
+                    content_item.render(out_fd, (indent_in + self.indent))
+                except AttributeError:
+                    out_fd.write('{0}{1}\n'.format(indent_in + self.indent,
+                                                   content_item))
+            out_fd.write('{0}</{1}>\n'.format(indent_in, self._tag))
         except IOError:
-            print("I/O Error with file [{0}] on open".format(out_file))
+            print("Element: I/O Error on render")
             return
+
+
+class OneLineTag(Element):
+    _attributes = {}
+
+    def append(self, content_in):
+        raise NotImplementedError
+
+    def render_attributes(self, out_fd, indent_in):
+        num_attrib = len(self._attributes)
+        if(num_attrib > 0):
+            out_fd.write('{0}<{1} '.format(indent_in, self._tag))
+            for key, value in self._attributes.items():
+                out_fd.write('{0}="{1}" '.format(key, value))
+        else:
+            out_fd.write('{0}<{1}>'.format(indent_in, self._tag))
+        return num_attrib
```

Comment:
You don't need to (and shouldn't) redefine `render_attributes()` here with the same duplicate code, because this class inherits from `Element`, where `render_attributes()` is already defined with exactly the same code. That's the whole point of inheritance: code reuse and avoidance of duplicate code.

Reply:
dup code issue resolved